### PR TITLE
[EN] Verify keyfiles signatures

### PR DESCRIPTION
**Pull request by cgrigis**
See #1307

### CODE REVIEW

**Comment by ale5000-git** [comment]
@cgrigis: Does it work also for [Corona Warn App](https://github.com/corona-warn-app/cwa-app-android)?

**Comment by cgrigis** [comment]
> @cgrigis: Does it work also for [Corona Warn App](https://github.com/corona-warn-app/cwa-app-android)?

@ale5000-git Not currently, because we don't have the public key from the German authorities. If anybody can obtain this, it will be trivial to add support. :)

**Comment by ale5000-git** [comment]
Corona Warn App support the transnational database that cover many European countries together, I'm interested in this rather than the German one.

**Comment by cgrigis** [comment]
Ah, my mistake. Nonetheless, we need whatever public key allows for their signature verification. Currently I only have the one from the Swiss PHA for SwissCovid — `ch.admin.bag.dp3t`.

**Comment by cgrigis** [comment]
@ale5000-git Do you know of a contact who could provide us with the public key for Corona Warn App?

**Comment by ale5000-git** [comment]
@cgrigis

Apparently there is
- `de.rki.coronawarnapp-dev` for testing
- `de.rki.coronawarnapp` for production

Some info here:
https://github.com/corona-warn-app/cwa-server/blob/main/docs/DISTRIBUTION.md
https://github.com/corona-warn-app/cwa-server/tree/main/docker-compose-test-secrets

This is a bit confusing for me, does it include what you need?

**Comment by cgrigis** [comment]
@ale5000-git 

Thank you for the pointers, that is useful information.

> https://github.com/corona-warn-app/cwa-server/tree/main/docker-compose-test-secrets
> 
> This is a bit confusing for me, does it include what you need?

Unfortunately (and understandably), the above contains the data for dev/test. What we need is the public key corresponding to the private key used by TSI to sign the key material sent by the production back-end...

File: play-services-nearby-core/src/main/kotlin/org/microg/gms/nearby/exposurenotification/ExposureNotificationServiceImpl.kt [hunk 1]
```diff
@@ -25,18 +26,53 @@ import org.json.JSONObject
...
+        } catch (e: Exception) {
+            Log.w(TAG, "Failed to retrieve back-end public key for ${packageName}: " + e.message)
+            null
+        }
+    }
+
+    // Table of supported signature algorithms for the diagnosed TEKs.
+    // The table is indexed by ASN.1 OIDs as specified in https://tools.ietf.org/html/rfc5758#section-3.2
+    private val sigAlgoForOid = mapOf<String, Signature>(
+            Pair("1.2.840.10045.4.3.2", Signature.getInstance("SHA256withECDSA")),
```
Comment (mar-v-in) [inline play-services-nearby-core/src/main/kotlin/org/microg/gms/nearby/exposurenotification/ExposureNotificationServiceImpl.kt:72]: Either map these to the name string (e.g. "SHA256withECDSA") and only run `Signature.getInstance()` when needed or map to a function that creates instances of Signature (by surrounding the invocation of `Signature.getInstance()` with curly braces).

PS: I very much like the notation of kotlin infix to for map creation, like
```kotlin
val sigAlgoForOid = mapOf<String, Function0<Signature>>(
  "1.2.840.10045.4.3.2" to { Signature.getInstance("SHA256withECDSA") },
  "1.2.840.10045.4.3.4" to { Signature.getInstance("SHA512withECDSA") }
)
```

Comment (cgrigis) [reply play-services-nearby-core/src/main/kotlin/org/microg/gms/nearby/exposurenotification/ExposureNotificationServiceImpl.kt]: I changed it as you suggested (indeed, the `... to ...` notation is quite nice, changed my other map accordingly :) ).

It is a trade-off though, exchanging the unused instances in memory for a new instantiation for every signature verification, right?

**Pull request merged**